code/modules/admin/species_ban.dm

```diff
@@ -0,0 +1,107 @@
+// Species ban system.
+// Disallows a player from using a specific species, but lets them play the game regardless
+// Useful if someone isnt playing a species properly at all (IE: Not following hard-driven species lore)
+
+/// Global list of all species bans. Key is a ckey, value is a list of species. Its a global list to avoid mass DB lookups.
+GLOBAL_LIST_EMPTY(species_bans_assoc)
+//GLOBAL_PROTECT(species_bans_assoc) // Mainly just to save space in the already bloated VV window
+
+/datum/admins/proc/add_species_ban(ckey, species)
+	if(!ckey || !species)
+		return
+	if(!GLOB.species_bans_assoc["[ckey]"]) // Incase someone has a ckey thats just all numbers
+		GLOB.species_bans_assoc["[ckey]"] = list()
+
+	GLOB.species_bans_assoc["[ckey]"] |= species // Only add once
+
+/datum/admins/proc/remove_species_ban(ckey, species)
+	if(!ckey || !species)
+		return
+	if(!GLOB.species_bans_assoc["[ckey]"]) // Incase someone has a ckey thats just all numbers
+		GLOB.species_bans_assoc["[ckey]"] = list()
+
+	GLOB.species_bans_assoc["[ckey]"] -= species
+
+// Global so it can be called from world/New()
+/proc/load_all_species_bans()
+	var/datum/db_query/select_query = SSdbcore.NewQuery("SELECT ckey, job FROM [format_table_name("ban")] WHERE bantype='SPECIES_BAN' AND isnull(unbanned)")
+
+	if(!select_query.warn_execute(async = FALSE)) // NOT async because its called in world/New()
+		qdel(select_query)
+		return FALSE
+
+	while(select_query.NextRow())
+		// Prep it all
+		var/banned_ckey = select_query.item[1]
+		var/banned_species = select_query.item[2]
+
+		if(!GLOB.species_bans_assoc["[banned_ckey]"]) // Incase someone has a ckey thats just all numbers
+			GLOB.species_bans_assoc["[banned_ckey]"] = list()
+
+		GLOB.species_bans_assoc["[banned_ckey]"] |= banned_species // Incase someone gets banned twice
+
+	qdel(select_query)
+
+// Global so it can be used in other places (IE: Conversion to another species)
+/proc/is_species_banned(ckey, species)
+	ASSERT(ckey)
+	ASSERT(species)
+	if(!GLOB.species_bans_assoc["[ckey]"]) // If they dont have an entry in the glob, they aint banned
+		return FALSE
+
+	// Cast to local list
+	var/list/species_cache = GLOB.species_bans_assoc["[ckey]"]
+	if(species in species_cache)
+		// If they in the list, they banned from it
+		return TRUE
+
+	// If we here, they aint banned
+	return FALSE
+
+
+/mob/verb/display_species_bans()
```

Comment:
Can't you use `GLOB.species_bans_assoc` instead of the DB here?

Reply:
Probably, but then I also implemented spam prevention for DB verbs randomly without it being used, but it still needs adding to avoid spamming "Check Job Bans", so I stuck with it this way. I feel like job bans is also done this way for a reason, even though they are checked with an assoc GLOB ingame as well.

Reply:
> I also implemented spam prevention for DB verbs

The DB cooldown system existing is not an argument for this solution

> I feel like job bans is also done this way for a reason

No, they're not. 😞 
![image](https://user-images.githubusercontent.com/15747030/105207115-6a59ae80-5b47-11eb-9640-9fcfb4e47b01.png)

Reply:
I mean, they are. I know this for a fact because I broke the job ban loading code on the live server and no one had their job bans because the assoc list was blank.